### ISSUE TITLE
Accessibility enhancements: add role=application to body. Add explicit (hidden) link to launch full view

### ISSUE THREAD

**Issue by karlgroves** [shower/shower#220]
Recent testing with a screen reader user has demonstrated two challenges I didn't anticipate in my earlier PR.

1. JAWS screen reader intercepts all keystrokes.  This can be alleviated by adding `role="application`" to the `<body>` element. This will enable JAWS users to traverse through the slides the same as other users. It also will have no negative impact on other users.

2. In order for screen reader users to present using Shower, they need to be able to launch the full slide view. Shower initially loads the slideshow in tiled format and users have to click on one of the slides in order to launch the full view. I think the easiest way to allow a screen reader user to launch the full view is to add a link (positioned offscreen) that allows the user to launch the first slide. 

@pepelsbey:  I'm interested in discussing how to handle these. Would you rather see these in HTML or would you want to see these parts added using JavaScript?  I see advantages to each approach. IMO adding this to Shower Core is more desirable so that we're 100% sure it is there for people who need it.

**Comment by LjWatson** [comment]
@KarlGroves thanks for filing this issue. A couple of thoughts...

The problem occurs for all Windows screen readers due to the interaction model they use. Background info here:
http://tink.uk/understanding-screen-reader-interaction-modes/

Adding the application role solves the problem of not being able to use the slide deck as intended. It also removes the ability to navigate and read slide content using standard (Windows) screen reader commands. Using the h shortcut to navigate between headings for example.

Would a "presentation mode" switch be viable perhaps? When in presentation mode the application role would be applied, enabling (Windows) screen reader users to present the slide deck using the JavaScript keyboard shortcuts provided by Shower, and when not in presentation mode to interact with the slide content in the usual way.

**Comment by dfarough** [comment]
I liked Leonie's  suggestion on this.  she suggested providing a Presentation mode switch in the content which would allow a user to change modes while in the presentation.  This would allow a screen reader user to switch modes in order to gain the benefits of using brows mode when the content required it and use the conventional mode so as to facilitate slide navigation and other Javascript provided functionality when necessary.

In my opinion, Application mode would certainly help, but I have found that getting out of application mode can be a problem at times. 

Incidentally, I think the article that Leonie provided is the best explanation of Screen reader interaction modes that I have seen.

**Comment by danhopkins** [comment]
I've proposed a (very simple) patch to just enable application role during slideshow. We don't seem to need an additional explicit control to enable slideshow, since that can be triggered by hitting return. It may be useful to document the available keyboard controls though.

**Comment by LjWatson** [comment]
@dfarough
It can be difficult to exit application mode, but in this instance I don't think that's a problem because whilst in slideshow the only interaction you really need is to operate the slides.

Thanks for your kind words about my post!

@danhopkins
This is a big step forward for screen reader usability with Shower. Thank you.

Documenting the keyboard shortcuts would definitely be a good thing. My initial suggestion to introduce a link/control was made because I didn't know a keyboard shortcut existed. It might also be helpful to include a note that explains Windows screen reader users will need to pass the shortcut back through to the browser for it to work - for example with Jaws it means pressing Jaws modifier 3, then the return key.

**Comment by pepelsbey** [comment]
@LjWatson, what would be the best way to include such note?

**Comment by LjWatson** [comment]
@pepelsbey
Probably something in the readme? Or where the readme mentions keyboard accessibility, linking through to a keyboard.md file perhaps? There aren't that many keyboard shortcuts though, so a single file just for that seems like it might be overkill though.

**Comment by pepelsbey** [comment]
@LjWatson, rewrote shortcuts docs from scratch, [how does it look](https://github.com/shower/shower/blob/master/docs/shortcuts-en.md)? I’ll put short version to readme and will link to the full one.

**Comment by LjWatson** [comment]
@pepelsbey This looks good to me. Thank you.

**Comment by fuzzbomb** [comment]
It's not clear how a keyboard-only user (sighted, without a screen reader) starts the slideshow.  @danhopkins earlier comment here says that hitting return starts the slideshow, but I can't get that to work.    Hitting return causes a blue border to appear around the first tile, and repeating this causes the blue border to jump to successive tiles.  I can't figure out how to start the slideshow.  I'm using version 2.5.1

**Comment by fuzzbomb** [comment]
I found shift-F5 and Cmd-Enter, but they're a bit obscure.  It's maybe OK for presenters to learn this control, but there's also the situation where someone posts their slides online for other people to use.  How would they discover shift-F5?

The docs for `Cmd Enter`, `Cmd Option P` and `Cmd Shift Enter` are Macintosh-sepcific.  I discovered that these work with the Windows key too, which could be made clearer.

**Comment by krigersergei** [comment]
@fuzzbomb I'm working on _a11y_ enhancements ([323](https://github.com/shower/shower/issues/323)), so also will take a look at what you said.
Any other ideas/comments are more than welcome.

**Comment by LJWatson** [comment]
Worth noting that the application role does not change the keyboard interaction for non AT users. It just makes the keyboard commands usable with Windows screen readers.

**Comment by danhopkins** [comment]
@fuzzbomb yep, unfortunately <return> support got broken in some update or other. win+f5 tends to work on linux, fwiw

**Comment by LJWatson** [comment]
AIRC the keyboard command for triggering slide view was changed a year or so back. It used to be enter, but now (on Windows at least) control f5 does the trick. Agree the discoverability of these is less than ideal.d

**Comment by krigersergei** [comment]
@LJWatson thanks a lot, I will add all these things in todo list of issues I found.

**Comment by pepelsbey** [comment]
[There’s dedicated page](https://github.com/shower/shower/blob/master/docs/shortcuts.md) in the docs with all available shortcuts listed. But I totally agree that discoverability could be better.

**Comment by pepelsbey** [comment]
> I found shift-F5 and Cmd-Enter in the docs, but they're a bit obscure

@fuzzbomb they weren’t just invented, I replicated shortcuts available in PowerPoint and Keynote to make it easier both for users of traditional presentation systems and compatible with remotes.

**Comment by pepelsbey** [comment]
As for Enter removal, I decided to make less exceptions in shortcuts. It used to be Enter for starting presentation and Enter and Shift Enter for moving forward and backward in presentation mode. Instead, I made Enter and Shift Enter work in boths mode as navigation shortcuts and Cmd Enter as dedicated shortcut for starting presentation.

Do you think it might help if I add Ctrl Enter as well?

Otherwise, there’s a plan to get rid of ARIA region in favor to making each slide focusable, so you would just Tab through them and activate them with Enter just like you’d do with links. It’s an early idea, but what you think?

**Comment by LJWatson** [comment]
>Otherwise, there’s a plan to get rid of ARIA region in favor to making each slide focusable, so you would just Tab through them and activate them with Enter just like you’d do with links. It’s an early idea, but what do you think?>

I think it would make things nearly unusable with a screen reader.

The way things work now, you enter slide view and can move between slides using enter/space/arrows/pgup/down (and these can also be mapped to bluetooth clickers).

Having to use tab and enter would be too much of a distraction when giving a talk. Removing the live region would also mean that a screen reader no longer automatically announced the slide title and/or content, so you'd additionally need to execute the screen reder commands for reading the slide content too.

**Comment by krigersergei** [comment]
@LJWatson we were talking about having focusable slides instead of hidden ARIA region, cause there is one issue with content duplication. When actively reading a content with a screen reader in full screen mode, the screen reader reads the content from both real slide and ARIA region.

Did you experienced this?

**Comment by LJWatson** [comment]
@krigersergei No, I don't think I have.

When in slide view (control f5) it isn't possible to read the slide content using standard screen reader commands (because of the application role). The only way is to move to the next/previous slide, and have the live region trigger the announcement.

If not in slide view, then the keyboard commands for moving to the previous/next slides don't work (because the application role hasn't been applied). The only way to read the slide content in this view, is by using standard screen reader commands.

Worth noting that you will get different behaviour on a Mac/VoiceOver. Windows screen readers have a different [keyboard intercept model](https://tink.uk/time-to-revisit-accesskey/) to Macs, which means that JavaScript driven shortcuts are handled differently on each platform (when a screen reader is running).

If you're using a Mac and get the results you describe, let me know. I'll dig mine out and test again. If you can let me know exactly what you're doing to get the problem, that'll help me replicate it.

**Comment by krigersergei** [comment]
@LJWatson Yes, I use MacOS (High Sierra)/VoiceOver for testing.

> If you can let me know exactly what you're doing to get the problem, that'll help me replicate it.

In short words VoiceOver reads more than we need when you are navigating through the slide. This issue is described in #323. Please let me know if it's not clear enough.

Since the current version of shower obviously is not working properly with screen readers and a keyboard, we are looking for some other ways of making it accessible. As @pepelsbey said focusable slides is just an idea, which we are currently testing. After testing it with VoiceOver I figured out that VoiceOver, for example, doesn't read the content immediately when focus is set dynamically with JavaScript, so maybe we need to figure out another way.

I would really appreciate if you as a "native" screen reader user give us some more detailed info about what keyboard and screen reader behavior you would like to have. I assume that your vision will a bit different from what I think. Thanks!

**Comment by krigersergei** [comment]
@pepelsbey @LJWatson 

I made some enhancements, please test now (check https://github.com/shower/core/pull/136 first).

Try to read slides in both *grid* and *slide*.

It's not a final implementation, just an idea how it can be, so any feedback is more than welcome.

So far I only tested in VoiceOver, more screen reader testing will be done a bit later.

**Comment by LJWatson** [comment]
@krigersergei do you have a working implementation that can be used for testing?

**Comment by krigersergei** [comment]
@LJWatson yes I do, in [this branch](https://github.com/shower/core/tree/accessible-slides).

Simply clone it, run `npm run serve` (should serve on port 8080 or whatever it prompts), then clone [basic presentation template](https://github.com/shower/shower) and replace `node_modules/shower-core/shower.min.js` in the bottom of `index.html` with `http://localhost:8080/shower.js` to use new code.

Also to get better screen reader experience you need to add following styles in `<head>`. Wothout them screen reader will not work properly. I already created 2 PRs for that, but they are still pending:

```css
.shower.full .slide {
    visibility: hidden;
}

.shower.full .slide.active {
    visibility: visible;
}

.shower.list .region {
    display: inherit;
}

.shower.full .region {
    display: inherit;
}
```

I tested it only with VoiceOver yet, will do full testing later.

If you have any questions please ping me.

**Comment by krigersergei** [comment]
@LJWatson I just deployed a built version [here](http://sergeikriger.com/shower-test/), hope that helps.

**Comment by LJWatson** [comment]
Thanks @krigersergei

I've been testing with Firefox 61 and Jaws 2018, and unfortunately there are lots of problems.

When in grid view, none of the slide content is available. Although each slide has an h2 and content, they are not accessible to the screen reader. This is because aria-hidden="true" has been applied to both the h2 and the content of each slide.

In addition, tabindex="-1" has been applied to each focusable element. This means that even if the content were accessible with a screen reader, it would not be possible to focus on or activate any of these interactive elements.

The purpose of grid view is to be able to see all the slides as a single document. The expectation is that the grid view can be navigated using standard web navigation commands (such as navigation by heading, tabbing through links, or using any of the many different reading commands).

Suggest removing all instances of aria-hidden="true" and tabindex="-1" when in grid view. I think these changes were made to fix a screen reader issue that isn't really valid.

When in grid view the message "Click shift + f5 to switch to slide mode" is repeated whenever a key is pressed. Similarly, when in slide view the message "You are in slide mode, click escape to exit" is repeated everytime you progress to a new slide.

These are useful messages, but it isn't necessary to hear them multiple times. I suggest using an alert when grid view is first loaded, and an alert when slide view is first loaded, so the user is told about the shortcut keys once.

The word "click" is perhaps not the best word to use in messages relating to keyboard shortcuts. I suggest using the word "use" instead. For example:
"Use shift + f5 to switch into slide view."

When in slide view the progress bar updates get in the way of the slide titles. This is a prblem with the existing version (but I don't use the progressbar and so I'd forgotten about it).

When I progress to a slide I hear something like "Slide progress 3.7 3.7 3.7 3.7 3.7%", before I hear the slide title.

The slide titles are also more complex than before, because the aria-label attribute has been used in order to provide more information ("slide title, slide 6 of 12" for example).

It is arguable whether this is helpful information or not. Does the engine take the slide title and apply the aria-label automatically, or is the author expected to do it?

When in slide view the slide content is no longer being announced by the screen reader. It is helpful to have the slide content spoken, in case you need a quick reminder of what you're suppsed to be saying!

Has the way the application role is applied changed? I've noticed a lot more problems where application mode isn't being triggered properly when the shortcut is used. It's possible this is because of Firefox, rather than because of changes to the engine though, I'm not sure right now.

Sorry this isn't more positive feedback. Thank you for putting time and energy into keeping Shower usable by screen reader users - I really do appreciate it.

**Comment by krigersergei** [comment]
@LJWatson short question: what do you expect screen reader to say when in grid mode you use arrow keys for selecting a slide (before you open it)?

**Comment by pepelsbey** [comment]
Let’s take a step back and think about the whole system again. Using built-in terminology there are two Shower modes: *list mode* (with all slides listed as thumbnails for a quick visual overview) and *full mode* (with current slide expanded fullscreen). The only difference between them are:

- `list` or `full` class name applied to the root element change styles.
- current slide fits viewport in full mode using scaled applied to the root element by JS.
- `next` elements are hidden in full mode until they are activated, but they are visible in list mode.

It’s possible to navigate between slides in both modes using the same shortcuts. So the difference is only visual, apart from `next` elements. HTML structure and behavior are the same. Is there an actual use or value in having list and full modes behaving differently for screen readers?

I wonder if the whole visual experience is worth recreating if content is already available.

**Comment by krigersergei** [comment]
@pepelsbey 

Either content readable in the list mode or not was just an idea taken from Keynote and PowerPoint and can be easily removed. After @LJWatson comments it's obvious that everything should be readable in all modes.

The problem in the previous solution is that content duplication didn't work well, so when you read the slide it's repeated twice. Since the current implementation didn't work well also, I will try to find another solution.

**Comment by LJWatson** [comment]
>It’s possible to navigate between slides in both modes using the same shortcuts. So the difference is only visual, apart from next elements. HTML structure and behavior are the same. Is there an actual use or value in having list and full modes behaving differently for screen readers?>

Ok, I think this is where we're getting tangled up. The list view shortcuts do work if you use VoiceOver (on the Mac), but they do not work if you use Jaws or NVDA (on Windows).

This is because screen readers on the different platforms behave differently. There is a detailed explanation in [Understanding screen reader interaction modes](https://tink.uk/understanding-screen-reader-interaction-modes/),but the short version is this:
* Jaws and NVDA intercept keystrokes before they reach the browser
* VoiceOver doesn't do this


So if you are in list view, and you use VoiceOver, pressing the arrow keys will execute the shortcut as @pepelsbey describes. If you use NVDA or Jaws though, the keystroke is intercepted before it gets as far as the JavaScript, and is instead used for a standard screen reader navigation command (moving between characters).


So we have two quite different modes of interaction happening in list view, depending on the platform/screen reader you're using.


When in slide view things behave as expected, no matter which screen reader you use. The application role causes Jaws and NVDA to stop intercepting the keystrokes, and instead passes them straight through to the browser where they execute the expected behaviour (moving between slides).


So the problem I think we need to solve is: how to get the slide content to be spoken by VoiceOver in list view, without breaking the way it works for Jaws and NVDA?

**Comment by LJWatson** [comment]
Looking at this more with help from @scottaohara, there might be a solution...

First of all, the problem of VoiceOver not reading the slides in list view, is only when using the arrow keys. VoiceOver does read the slides when using standard web navigation commands.

The problem seems to be caused by VoiceOver not tracking focus properly, when the arrow keys are used to highlight the next slide in the set. Not sure why this should be though.

One possible solution is to put tabindex="0" on the ```<section>``` for each slide, when in list view. It's a completely dirty hack, but some quick tests with @scottaohara sugest it solves the problem.

**Comment by krigersergei** [comment]
@LJWatson to put `tabindex="0"` was my first idea. I tried it out and got some problems, such as not all screen readers read dynamically focused element in the same way. Some of them read the whole element content, some of them read nothing. For now I agree with @scottaohara that it may solve the problem and will implement it.

**Comment by LJWatson** [comment]
> to put tabindex="0" was my first idea. I tried it out and got some problems, such as not all screen readers read dynamically focused element in the same way. Some of them read the whole element content, some of them read nothing. For now I agree with @scottaohara that it may solve the problem and will implement it.>

In conversation with @stevefaulkner, he suggested explicitly putting a role on the ```<section>``` element, since this may help.
```
<section role="region" aria-label="...">...</section>
```

I think if you can do this, and take out all the aria-hidden attributes, we might be close to a solution. That works with all screen readers, if a little differently for each one.

**Comment by krigersergei** [comment]
@LJWatson thanks, will try it out.

**Comment by krigersergei** [comment]
@LJWatson I made some changes @stevefaulkner suggested, could you please check it? Here is a [deployed version](http://sergeikriger.com/shower-test/) of last changes or you can pull it from `shower/core`.

In VoiceOver Chrome it was quite nice with only thing that VoiceOver didn't read both ordered and unordered lists automatically, so I had to read them with keyboard.

In NVDA and JAWS it wasn't that good. But maybe I'm using these screen readers in a wrong way, so your feedback needed.

**Comment by krigersergei** [comment]
@scottaohara each slide gets `role=region` dynamically once it's active. `aria-label` is not used, because this way screen readers (at least VoiceOver) read only a label, but the goal is that screen readers read all the content.
